server/src/main/java/org/jboss/as/server/mgmt/ManagementWorkerService.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * JBoss, Home of Professional Open Source
+ * Copyright 2016, Red Hat, Inc., and individual contributors as indicated
+ * by the @authors tag.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jboss.as.server.mgmt;
+
+import java.io.IOException;
+
+import org.jboss.msc.service.Service;
+import org.jboss.msc.service.ServiceController;
+import org.jboss.msc.service.ServiceName;
+import org.jboss.msc.service.ServiceTarget;
+import org.jboss.msc.service.StartContext;
+import org.jboss.msc.service.StartException;
+import org.jboss.msc.service.StopContext;
+import org.xnio.OptionMap;
+import org.xnio.Options;
+import org.xnio.Xnio;
+import org.xnio.XnioWorker;
+
+/**
+ * @author Tomaz Cerar (c) 2016 Red Hat Inc.
+ */
+public class ManagementWorkerService implements Service<XnioWorker> {
+    public static final ServiceName SERVICE_NAME = ServiceName.JBOSS.append("serverManagement", "controller", "management", "worker");
+
+
+    private final OptionMap options;
+    private volatile XnioWorker worker;
+
+    private ManagementWorkerService(OptionMap options) {
+        this.options = options;
+    }
+
+    @Override
+    public void start(StartContext startContext) throws StartException {
+        final Xnio xnio = Xnio.getInstance();
+
+        try {
+            worker = xnio.createWorker(options);
+        } catch (IOException e) {
+            throw new StartException("Could not create worker!", e);
+        }
+    }
+
+    @Override
+    public void stop(StopContext stopContext) {
+        worker.shutdown();
```

Comment:
This is wrong, instead you should do this:

```
    private final AtomicReference<StopContext> stopRef = new AtomicReference<>();
    // ...
        public void start(final StartContext context) throws StartException {
        final Xnio xnio = Xnio.getInstance();
        try {
            xnioWorker = xnio.createWorker(null, options, () -> stopDone());
        } catch (IOException e) {
            throw new StartException(e);
        }
    }

    public void stop(final StopContext context) {
        stopRef.set(context);
        context.asynchronous();
        xnioWorker.shutdown();
        xnioWorker = null;
    }

    private void stopDone() {
        final StopContext stopContext = stopRef.get();
        assert stopContext != null;
        stopContext.complete();
    }
    // ...
```

Otherwise the service will stop before the worker is stopped.